tools/http-fault-injector/sample-clients/java/storage-blobs/src/main/java/Utils.java

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import com.azure.core.http.HttpRequest;
+
+import java.net.URI;
+import java.util.Objects;
+
+/**
+ * Utility class containing constants and methods for re-writing {@link HttpRequest HttpRequests} to use the HTTP fault
+ * injector.
+ */
+public final class Utils {
+    /**
+     * The default host used by HTTP fault injector.
+     */
+    public static final String DEFAULT_HTTP_FAULT_INJECTOR_HOST = "localhost";
+
+    /**
+     * The default HTTP port used by HTTP fault injector.
+     */
+    public static final int DEFAULT_HTTP_FAULT_INJECTOR_HTTP_PORT = 7777;
+
+    /**
+     * The default HTTPS port used by HTTP fault injector.
+     */
+    public static final int DEFAULT_HTTP_FAULT_INJECTOR_HTTPS_PORT = 7778;
+
+    /**
+     * The HTTP header used by HTTP fault injector to determine where it needs to forward a request.
+     */
+    public static final String HTTP_FAULT_INJECTOR_UPSTREAM_HOST_HEADER =  "X-Upstream-Host";
+
+    /**
+     * Utility method which re-writes the {@link HttpRequest HttpRequest's} URL to use the HTTP fault injector.
+     * <p>
+     * This will set the HTTP header {@link #HTTP_FAULT_INJECTOR_UPSTREAM_HOST_HEADER} to the request URL used by the
+     * HTTP request before re-writing and will update the request URL to use the HTTP fault injector.
+     *
+     * @param request The {@link HttpRequest} having its URL re-written.
+     * @param host The HTTP fault injector host.
+     * @param port The HTTP fault injector port.
+     * @return The updated {@link HttpRequest} with its URL re-written.
+     * @throws NullPointerException If {@code request} or {@code host} are null.
+     * @throws IllegalArgumentException If {@code host} is an empty string or {@code port} is
+     * an invalid port.
+     * @throws IllegalStateException If the request URL isn't valid or the HTTP fault injector URL isn't valid.
+     */
+    public static HttpRequest rewriteUrlToUseFaultInjector(HttpRequest request, String host, int port) {
+        validateHostAndPort(host, port);
+
+        try {
+            URI requestUri = request.getUrl().toURI();
+            URI faultInjectorUri = new URI(requestUri.getScheme(), requestUri.getUserInfo(), host,
+                port, requestUri.getPath(), requestUri.getQuery(), requestUri.getFragment());
+
+            String xUpstreamHost = (requestUri.getPort() < 0)
```

Comment:
In .NET, the `HttpMessage` is reused on retries, so we need to ensure the header is only set the first time.  Overwriting the header is incorrect, since the host and port will point to the fault injector on the retry, so if `X-Upstream-Host` is set again it will create an infinite loop (both the request URL and the `X-Upstream-Host` will point to the fault injector).

Reply:
Test with retry scenarios in Java, we [clone deeply per retry](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RetryPolicy.java#L96) ([HttpReqeust.copy](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core/src/main/java/com/azure/core/http/HttpRequest.java#L214) and [HttpHeaders](https://github.com/Azure/azure-sdk-for-java/blob/1e670b28aa53f13ff5b15538d320118d38a5e299/sdk/core/azure-core/src/main/java/com/azure/core/http/HttpHeaders.java#L42))